### DISCUSSION TITLE
Why do vllm set default keep-alive timeout to 5s?

### DISCUSSION THREAD

**Discussion by SeungminHeo** [vllm-project/vllm#8287] [Q&A]
Hi,

For huge size and long-context models takes more than 5s and no additional connections or requests are sent to server, so server can close connection and trigger exception. For example, openai sdk that uses httpx makes error message`Server disconnected without sending a response.`

For this, I want to control this with uvicorn arguments, can this be overriden?
This is a "Guess", but exceptions are actually happening.

**Comment by aidenhe** [comment]
now, you can set env change this value 
see  https://github.com/vllm-project/vllm/pull/18472

**Comment by gitsult4n** [comment]
5s matches uvicorn's own default
`uvicorn/config.py:229` -> `timeout_keep_alive: int = 5`
vllm sets the same value at `vllm/envs.py:109`
i found no commit or comment stating a reason for that particular number

it is an idle timeout and not a request timeout so it cannot kill a long generation
the timer is armed only after a response completes (`h11_impl.py:322-331`) and any incoming data cancels it (`h11_impl.py:172-173`)
i ran uvicorn `0.52.4` with `timeout_keep_alive=2` behind a handler that sleeps 6s and it still returned 200 after 6s

what does break is writing onto a pooled connection the server already closed
the idle handler closes it (`h11_impl.py:364-371`) and httpcore turns the resulting EOF into your exact message (`httpcore/_async/http11.py:229-231`)
i reproduced `Server disconnected without sending a response.` against a server that answers the first request then FINs the second
guess -> plain idle reuse often does not raise since httpcore checks the connection before reusing it so a proxy or a dead engine can look the same

there is no python cli flag for it so use the env var

```
VLLM_HTTP_TIMEOUT_KEEP_ALIVE=300 vllm serve <model>
```

defined at `vllm/envs.py:1115-1117` and consumed at `vllm/entrypoints/launchers/api_server/entry.py:151`
needs `v0.9.1` or newer since #18472 landed after `v0.9.0`
before that it was a hardcoded `TIMEOUT_KEEP_ALIVE = 5` with no override
the rust frontend takes `--http-timeout-keep-alive` and reads the same env var (`rust/src/cmd/src/cli.rs:239-241`)
